src/tests/string/test_strings_format_string.f90

```diff
@@ -0,0 +1,82 @@
+program test_strings_format_string
+    use stdlib_strings, only: format_string, starts_with
+    use stdlib_error, only: check
+    use stdlib_optval, only: optval
+    implicit none
+    print *, 'format_string(complex) : '
+    call check_formatter(format_string((1, 1)), "(1.0", &
+        & "Default formatter for complex number", partial=.true.)
+    call check_formatter(format_string((1, 1), '(F6.2)'), "(  1.00,  1.00)", &
+        & "Formatter for complex number")
+    call check_formatter(format_string((-1, -1), '(F6.2)'), "( -1.00, -1.00)", &
+        & "Formatter for negative complex number")
+    call check_formatter(format_string((1, 1), '(SP,F6.2)'), "( +1.00, +1.00)", &
+        & "Formatter with sign control descriptor for complex number")
+    call check_formatter(format_string((1, 1), '(F6.2)')//format_string((2, 2), '(F7.3)'), &
+        & "(  1.00,  1.00)(  2.000,  2.000)", &
+        & "Multiple formatters for complex numbers")
+    print *, 'format_string(integer) : '
+    call check_formatter(format_string(100), "100", &
+        & "Default formatter for integer number")
+    call check_formatter(format_string(100, '(I6)'), "   100", &
+        & "Formatter for integer number")
+    call check_formatter(format_string(100, '(I0.6)'), "000100", &
+        & "Formatter with zero padding for integer number")
+    call check_formatter(format_string(100, '(I6)')//format_string(1000, '(I7)'), &
+        & "   100   1000", &
+        & "Multiple formatters for integers")
+    call check_formatter(format_string(34, '(B8)'), "  100010", &
+        & "Binary formatter for integer number")
+    call check_formatter(format_string(34, '(O0.3)'), "042", &
+        & "Octal formatter with zero padding for integer number")
+    call check_formatter(format_string(34, '(Z3)'), " 22", &
+        & "Hexadecimal formatter for integer number")
+    print *, 'format_string(real) : '
+    call check_formatter(format_string(100.), "100.0", &
+        & "Default formatter for real number", partial=.true.)
+    call check_formatter(format_string(100., '(F6.2)'), "100.00", &
+        & "Formatter for real number")
+    call check_formatter(format_string(289., '(E7.2)'), ".29E+03", &
+        & "Exponential formatter with rounding for real number")
+    call check_formatter(format_string(128., '(ES8.2)'), "1.28E+02", &
+        & "Exponential formatter for real number")
+    ! Wrong demonstration
+    call check_formatter(format_string(-100., '(F6.2)'), "*", &
+        & "Too narrow formatter for signed real number", partial=.true.)
+    call check_formatter(format_string(1000., '(F6.3)'), "*", &
+        & "Too narrow formatter for real number", partial=.true.)
+    call check_formatter(format_string(1000, '(F7.3)'), "*", &
+        & "Real formatter for integer number", partial=.true.)
```

Comment:
`ifort` seems to implicitly cast the integer into real before printing, which is smart but not consistent with `gfortran`:

```
 Real formatter for integer number
Expected: '*' but got '  0.000'
```

I'm not sure how to make a test that consistently fails here, maybe by passing an invalid format descriptor like here:

```suggestion
    call check_formatter(format_string(1000., '(7.3)'), "*", &
        & "Invalid formatter for real number", partial=.true.)
```

Reply:
Sorry i just saw this comment.

Reply:
```fortran
read(str, '(7.3)', iostat=stat) 1000
```
The above statement is on `gfortran`, will product `stat\=0`, but on `ifort`, will product `stat=0`, and converts `1000` to `    0`(see [here](https://github.com/fortran-lang/stdlib/runs/2920528642)). This situation also occurs in the following conditions.
![image](https://user-images.githubusercontent.com/32035096/123532317-c05a0000-d73e-11eb-8d86-9e0463a2eb69.png)

I don't think this is the flexibility or smart of `ifort`(see https://github.com/fortran-lang/stdlib/issues/446). In these cases, `ifort` should have reported errors to users, but it has been transformed, and the transformed results seem to lack practical significance.